Refactor remote function calls to implement typesafe interfaces

cdharris opened the pull request:
Planning the refactor of remote functions to be type safe as per https://github.com/WorldBrain/Memex/issues/836

cdharris commented:
I found it tricky trying to design a proxy along the lines of the original suggestion that would allow both the concrete class to implement and also allow the proxying function/class to return the interface too, the design in https://github.com/WorldBrain/Memex/pull/837/commits/3f96be92982a067ec1ac609cdf89c80e70b27700 was what seemed to work best for me. Though perhaps there's something more inline with the original thinking.

Some Results of digging into this issue further..

## Current usage audit of `makeRemotelyCallable`

- There's 112 of these functions!
- I tried to count their usages, but they are used in quite different ways (used directly by their string name, some assigned to variables first, some with names called by dynamic variables)
- One function is registered twice (see below)
- I can't tell if they are all definitely used

### `trackEvent`
RPC Function `trackEvent` is defined in `src/analytics/background/index.js:10` which uses `=> analytics.trackEvent` from the `const analytics: Analytics = new AnalyticsManager({ backend })
` exported at `src/analytics/index.ts:24`

but `trackEvent` is also defined in `src/analytics/internal/background/index.js:11` which uses `=> sendToServer.trackEvent` from `src/analytics/internal/send-to-server/send-to-server.js:109`

This latter definition seems to be registered last so seems to be the one that will be used.

Q: What is the difference between `src/analytics/...` and `src/analytics/internal/...` anyway ?


### Appendix

<details>
<summary>RPC Function registration order</summary>
```
createNotification
isURLBlacklisted
addToBlacklist
trackEvent
updateLastActive
dirtyEstsCache
storeNotification
fetchUnreadCount
fetchUnreadNotifications
fetchReadNotifications
readNotification
fetchNotifById
dispatchNotification
addTweet
addPostToList
delPostFromList
fetchSocialPostLists
delSocialPages
addSocialBookmark
delSocialBookmark
addTagForTweet
delTagForTweet
fetchSocialPostTags
fetchUserSuggestions
fetchAllUsers
fetchAllHashtags
fetchHashtagSuggestions
createDirectLink
getAllAnnotationsByUrl
createAnnotation
editAnnotation
editAnnotationTags
deleteAnnotation
getAnnotationTags
addAnnotationTag
delAnnotationTag
followAnnotationRequest
toggleSidebarOverlay
toggleAnnotBookmark
insertAnnotToList
removeAnnotFromList
goToAnnotationFromSidebar
toggleLoggingPause
fetchTab
fetchTabByUrl
search
addPageTag
delPageTag
suggest
delPages
addPageBookmark
delPageBookmark
fetchPageTags
extendedSuggest
delPagesByDomain
delPagesByPattern
getMatchingPageCount
searchAnnotations
searchPages
searchSocial
storeEvent
getLatestTimeWithCount
trackEvent
processEvent
createCustomList
insertPageToList
updateListName
removeList
removePageFromList
fetchAllLists
fetchListById
fetchListPagesByUrl
fetchListNameSuggestions
fetchListPagesById
fetchListIgnoreCase
addOpenTabsToList
removeOpenTabsFromList
addTag
delTag
addTagsToOpenTabs
delTagsFromOpenTabs
addBookmark
delBookmark
getBackupProviderLoginLink
startBackup
initRestoreProcedure
getBackupInfo
pauseBackup
resumeBackup
cancelBackup
startRestore
getRestoreInfo
pauseRestore
resumeRestore
cancelRestore
hasInitialBackup
setBackendLocation
getBackendLocation
isBackupAuthenticated
maybeCheckAutomaticBakupEnabled
checkAutomaticBakupEnabled
isAutomaticBackupEnabled
sendNotification
estimateInitialBackupSize
setBackupBlobs
getBackupTimes
forgetAllChanges
storeWordpressUserId
setupRequestInterceptor
openOverviewTab
openOptionsTab
openLearnMoreTab
```

</details>


<details>
<summary>All Code usages of setupRemoteFunctions</summary>

`src/activity-logger/background/index.ts:62`
```typescript
    setupRemoteFunctions() {
        makeRemotelyCallable({
            toggleLoggingPause: this.toggleLoggingPause,
            fetchTab: id => this.tabManager.getTabState(id),
            fetchTabByUrl: url => this.tabManager.getTabStateByUrl(url),
        })
    }
```

`src/analytics/background/index.js:10`
```typescript
 makeRemotelyCallable({
    trackEvent: (...args) => analytics.trackEvent(...args),
    updateLastActive,
})
```


`src/analytics/internal/background/index.js:11`
```typescript
    setupRemoteFunctions() {
        makeRemotelyCallable({
            storeEvent: (...params) => {
                return this.storeEvent(...params)
            },
            getLatestTimeWithCount: (...params) => {
                return this.getLatestTimeWithCount(...params)
            },
            trackEvent: (...params) => {
                return sendToServer.trackEvent(...params)
            },
            processEvent: (...params) => {
                return internalAnalytics.processEvent(...params)
            },
        })
    }
```

`src/background-script/index.ts:154`
```typescript
    setupRemoteFunctions() {
        makeRemotelyCallable({
            openOverviewTab: this.utils.openOverviewURL,
            openOptionsTab: this.utils.openOptionsURL,
            openLearnMoreTab: this.utils.openLearnMoreURL,
        })
    }
```


`src/backup/background/index.ts:61`
```typescript
       setupRemoteFunctions() {
        makeRemotelyCallable(
            {
                getBackupProviderLoginLink: async (info, params) => {
                    const MEMEX_CLOUD_ORIGIN = _getMemexCloudOrigin()
                    return `${MEMEX_CLOUD_ORIGIN}/auth/google?scope=${DEFAULT_AUTH_SCOPE}`
                },
                startBackup: ({ tab }, params) => {
                    this.backupUiCommunication.registerUiTab(tab)
                    if (this.backupProcedure.running) {
                        return
                    }
                    if (
                        this.restoreProcedure &&
                        this.restoreProcedure.running
                    ) {
                        throw new Error(
                            "Come on, don't be crazy and run backup and restore at once please",
                        )
                    }

                    this.doBackup()
                    this.backupUiCommunication.connect(
                        this.backupProcedure.events,
                    )
                },
                initRestoreProcedure: (info, provider) => {
                    return this.initRestoreProcedure(provider)
                },
                getBackupInfo: () => {
                    return this.backupProcedure.info
                },
                pauseBackup: () => {
                    this.backupProcedure.pause()
                },
                resumeBackup: () => {
                    this.backupProcedure.resume()
                },
                cancelBackup: async () => {
                    await this.backupProcedure.cancel()
                },
                startRestore: async ({ tab }) => {
                    this.restoreUiCommunication.registerUiTab(tab)
                    await this.startRestore()
                },
                getRestoreInfo: async () => {
                    return this.restoreProcedure.info
                },
                pauseRestore: async () => {
                    await this.restoreProcedure.interruptable.pause()
                },
                resumeRestore: async () => {
                    await this.restoreProcedure.interruptable.resume()
                },
                cancelRestore: async () => {
                    await this.restoreProcedure.interruptable.cancel()
                },
                hasInitialBackup: async () => {
                    return !!(await this.lastBackupStorage.getLastBackupTime())
                },
                setBackendLocation: async (info, location?: string) => {
                    if (
                        location === 'google-drive' &&
                        this.backendLocation !== location
                    ) {
                        this.backendLocation = location
                        await this.backendSelect.saveBackendLocation(location)
                        this.backend = await this.backendSelect.initGDriveBackend()
                    } else if (
                        location === 'local' &&
                        this.backendLocation !== location
                    ) {
                        this.backendLocation = location
                        await this.backendSelect.saveBackendLocation(location)
                        this.backend = await this.backendSelect.initLocalBackend()
                    }
                    this.setupRequestInterceptor()
                    this.initBackendDependants()
                },
                getBackendLocation: async info => {
                    this.backendLocation = await this.backendSelect.restoreBackendLocation()
                    return this.backendLocation
                },
                isBackupAuthenticated: async () => {
                    let backend = null
                    /* Check if restoreProcedure's backend is present.
                        Restore's backend is only present during restore. */
                    backend = this.restoreProcedure
                        ? this.restoreProcedure.backend
                        : this.backend

                    return backend ? backend.isAuthenticated() : false
                },
                maybeCheckAutomaticBakupEnabled: async () => {
                    if (
                        !!(await this.lastBackupStorage.getLastBackupTime()) &&
                        localStorage.getItem('wp.user-id') &&
                        localStorage.getItem('backup.has-subscription') &&
                        localStorage.getItem('nextBackup') === null
                    ) {
                        await this.checkAutomaticBakupEnabled()
                        await this.maybeScheduleAutomaticBackup()
                    }
                },
                checkAutomaticBakupEnabled: async () => {
                    // The only place this is called right now is post-purchase.
                    // Move to more suitable place once this changes.
                    const override =
                        process.env.AUTOMATIC_BACKUP_PAYMENT_SUCCESS
                    if (override && override.length) {
                        console.log(
                            'Automatic backup payment override',
                            override,
                        )
                        this.automaticBackupCheck = Promise.resolve(
                            override === 'true',
                        )
                        // Send a notification stating that the auto backup has expired
                        this.notifications.dispatchNotification(
                            'auto_backup_expired',
                        )
                        // Set the message of backupStatus stating the expiration of auto backup
                        await setLocalStorage('backup-status', {
                            state: 'fail',
                            backupId: 'auto_backup_expired',
                        })
                    } else {
                        await this.checkAutomaticBakupEnabled()
                    }

                    return this.automaticBackupCheck
                },
                isAutomaticBackupEnabled: async () => {
                    return this.isAutomaticBackupEnabled()
                },
                sendNotification: async (id: string) => {
                    const errorId = await this.backend.sendNotificationOnFailure(
                        id,
                        this.notifications,
                        () => this.estimateInitialBackupSize(),
                    )
                    return errorId
                },
                estimateInitialBackupSize: () => {
                    return this.estimateInitialBackupSize()
                },
                setBackupBlobs: (info, saveBlobs) => {
                    localStorage.setItem('backup.save-blobs', saveBlobs)
                },
                getBackupTimes: async () => {
                    return this.getBackupTimes()
                },
                forgetAllChanges: async () => {
                    return this.forgetAllChanges()
                },
                storeWordpressUserId: (info, userId) => {
                    localStorage.setItem('wp.user-id', userId)
                },
                setupRequestInterceptor: () => {
                    return this.setupRequestInterceptor()
                },
            },
            { insertExtraArg: true },
        )
    }
```

`makeRemotelyCallable`
```typescript
    makeRemotelyCallable({
        isURLBlacklisted: blacklist.isURLBlacklisted,
        addToBlacklist: blacklist.addToBlacklist,
})

```


`src/bookmarks/background/index.ts:13`
```typescript
        setupRemoteFunctions() {
        makeRemotelyCallable({
            addBookmark: this.addBookmark.bind(this),
            delBookmark: this.delBookmark.bind(this),
        })
    }
```


`src/content-tooltip/interactions.js:130`
```typescript
   export const setupRPC = ({ toolbarNotifications }) => {
    makeRemotelyCallable({
        showContentTooltip: async () => {
            if (!showTooltip) {
                await insertTooltip({ toolbarNotifications })
            }
            if (userSelectedText()) {
                const position = calculateTooltipPostion()
                showTooltip(position)
            }
        },
        insertTooltip: ({ override } = {}) => {
            manualOverride = !!override
            insertTooltip({ toolbarNotifications })
        },
        removeTooltip: ({ override } = {}) => {
            manualOverride = !!override
            removeTooltip()
        },
        insertOrRemoveTooltip: async () => {
            await insertOrRemoveTooltip({ toolbarNotifications })
        },
    })
}
```

`src/custom-lists/background/index.ts:39`
```typescript
        setupRemoteFunctions() {
        makeRemotelyCallable({
            createCustomList: this.createCustomList.bind(this),
            insertPageToList: this.insertPageToList.bind(this),
            updateListName: this.updateList.bind(this),
            removeList: this.removeList.bind(this),
            removePageFromList: this.removePageFromList.bind(this),
            fetchAllLists: this.fetchAllLists.bind(this),
            fetchListById: this.fetchListById.bind(this),
            fetchListPagesByUrl: this.fetchListPagesByUrl.bind(this),
            fetchListNameSuggestions: this.fetchListNameSuggestions.bind(this),
            fetchListPagesById: this.fetchListPagesById.bind(this),
            fetchListIgnoreCase: this.fetchListIgnoreCase.bind(this),
            addOpenTabsToList: this.addOpenTabsToList.bind(this),
            removeOpenTabsFromList: this.removeOpenTabsFromList.bind(this),
        })
    }
```

`src/direct-linking/background/index.ts:55`
```typescript
    setupRemoteFunctions() {
        makeRemotelyCallable(
            {
                createDirectLink: this.createDirectLink.bind(this),
                getAllAnnotationsByUrl: this.getAllAnnotationsByUrl.bind(this),
                createAnnotation: this.createAnnotation.bind(this),
                editAnnotation: this.editAnnotation.bind(this),
                editAnnotationTags: this.editAnnotationTags.bind(this),
                deleteAnnotation: this.deleteAnnotation.bind(this),
                getAnnotationTags: this.getTagsByAnnotationUrl.bind(this),
                addAnnotationTag: this.addTagForAnnotation.bind(this),
                delAnnotationTag: this.delTagForAnnotation.bind(this),
                followAnnotationRequest: this.followAnnotationRequest.bind(
                    this,
                ),
                toggleSidebarOverlay: this.toggleSidebarOverlay.bind(this),
                toggleAnnotBookmark: this.toggleAnnotBookmark.bind(this),
                insertAnnotToList: this.insertAnnotToList.bind(this),
                removeAnnotFromList: this.removeAnnotFromList.bind(this),
                goToAnnotationFromSidebar: this.goToAnnotationFromSidebar.bind(
                    this,
                ),
            },
            { insertExtraArg: true },
        )
    }
```

`src/imports/background/index.js:12`
```typescript
    makeRemotelyCallable({
    dirtyEstsCache: () => importStateManager.dirtyEstsCache(),
})
```

`src/notifications/background/index.js:22`
```typescript
       setupRemoteFunctions() {
        makeRemotelyCallable({
            storeNotification: (...params) => {
                return this.storeNotification(...params)
            },
            fetchUnreadCount: () => {
                return this.fetchUnreadCount()
            },
            fetchUnreadNotifications: () => {
                return this.fetchUnreadNotifications()
            },
            fetchReadNotifications: (...params) => {
                return this.fetchReadNotifications(...params)
            },
            readNotification: id => {
                return this.readNotification(id)
            },
            fetchNotifById: id => {
                return this.fetchNotifById(id)
            },
            dispatchNotification: notification => {
                return this.dispatchNotification(notification)
            },
        })
    }
```

`src/page-analysis/content_script/index.js:6`
```typescript
    makeRemotelyCallable({
    extractPageContent,
})
```

`src/search/background/index.ts:112`
```typescript
   setupRemoteFunctions() {
        makeRemotelyCallable({
            search: this.backend.search,
            addPageTag: this.backend.addTag,
            delPageTag: this.backend.delTag,
            suggest: this.backend.suggest,
            delPages: this.backend.delPages,
            addPageBookmark: this.backend.addBookmark,
            delPageBookmark: this.backend.delBookmark,
            fetchPageTags: this.backend.fetchPageTags,
            extendedSuggest: this.backend.extendedSuggest,
            delPagesByDomain: this.backend.delPagesByDomain,
            delPagesByPattern: this.backend.delPagesByPattern,
            getMatchingPageCount: this.backend.getMatchingPageCount,
            searchAnnotations: this.searchAnnotations.bind(this),
            searchPages: this.searchPages.bind(this),
            searchSocial: this.searchSocial.bind(this),
        })
    }
```

`src/search/migration/index.ts:16`
```typescript
    makeRemotelyCallable({
    startMigration: async () => {
        isRunning = true
        await migrator.start()
        MigrationManager.showNotif()
    },
    isMigrating: () => isRunning,
})
```

`src/sidebar-overlay/content_script/ribbon-interactions.ts:173`

```typescript
    export const setupRPC = ({
    annotationsManager,
    toolbarNotifications,
}: {
    annotationsManager: AnnotationsManager
    toolbarNotifications: ToolbarNotifications
}) => {
    makeRemotelyCallable({
        /**
         * Used for inserting the ribbon.
         */
        insertRibbon: (
            { override, ...args }: { override: boolean } = {
                override: false,
            },
        ) => {
            manualOverride = !!override
            insertRibbon({
                annotationsManager,
                toolbarNotifications,
                ...args,
            })
        },
        /**
         * Used for removing the ribbon.
         */
        removeRibbon: ({ override } = { override: false }) => {
            manualOverride = !!override
            removeRibbon()
        },
        /**
         * Inserts or removes the ribbon from the page depending on whether it
         * is enabled or not.
         */
        insertOrRemoveRibbon: async () => {
            await _insertOrRemoveRibbon({
                annotationsManager,
                toolbarNotifications,
            })
        },
        /**
         * RPC for updating the ribbon's state.
         */
        updateRibbon,
    })
}
```


`src/sidebar-overlay/ribbon-sidebar-controller/ribbon-sidebar-container.tsx:179`
```typescript
   private _setupRPC = () => {
        makeRemotelyCallable({
            openSidebar: this._openSidebar,
            goToAnnotation: this._goToAnnotation,
        })
    }

```



`src/social-integration/background/index.ts:17`
```typescript
       setupRemoteFunctions() {
        makeRemotelyCallable({
            addTweet: this.addTweet.bind(this),
            addPostToList: this.addPostToList.bind(this),
            delPostFromList: this.delPostFromList.bind(this),
            fetchSocialPostLists: this.fetchSocialPostLists.bind(this),
            delSocialPages: this.delSocialPages.bind(this),
            addSocialBookmark: this.addSocialBookmark.bind(this),
            delSocialBookmark: this.delSocialBookmark.bind(this),
            addTagForTweet: this.addTagForTweet.bind(this),
            delTagForTweet: this.delTagForTweet.bind(this),
            fetchSocialPostTags: this.fetchSocialPostTags.bind(this),
            fetchUserSuggestions: this.fetchUserSuggestions.bind(this),
            fetchAllUsers: this.fetchAllUsers.bind(this),
            fetchAllHashtags: this.fetchAllHashtags.bind(this),
            fetchHashtagSuggestions: this.fetchHashtagSuggestions.bind(this),
        })
    }

```



`src/tags/background/index.ts:38`
```typescript
    setupRemoteFunctions() {
        makeRemotelyCallable({
            addTag: this.addTag.bind(this),
            delTag: this.delTag.bind(this),
            addTagsToOpenTabs: this.addTagsToOpenTabs.bind(this),
            delTagsFromOpenTabs: this.delTagsFromOpenTabs.bind(this),
        })
    }
```



`src/util/notifications.ts:56`
```typescript
    makeRemotelyCallable({ createNotification })
```



`src/toolbar-notification/content_script/index.ts:10`
```typescript
   registerRemoteFunctions(registry: RemoteFunctionRegistry) {
        registry.registerRemotelyCallable({
            showToolbarNotification: (type: string, extraProps: any) => {
                this.showToolbarNotification(type, extraProps)
            },
        })
    }
```
(note this last case uses `RemoteFunctionRegistry.registerRemotelyCallable` which is only used once and forwards to `makeRemotelyCallable)
</details>


This is looking like a considerable refactor job

cdharris commented:
Just to re-iterate a little bit clearer the important part of I've done here in https://github.com/WorldBrain/Memex/pull/837/commits/db5bcb9eeb0c6d6317d750e46e825c189a75ba19 

- started with the remote function `createNotification` 
- wrapped it in a class to improve organisation and method typehinting from the interface it implements
- changed client usages of 
```typescript
this.createNotif = remoteFunction('createNotification')
```
   to
```typescript
this.createNotif = remoteInterface<NotificationInterface>().createNotification
``` 


Optionally, I could change the client usages a bit further from what they are at the moment to better suit this new interface by doing something like
```typescript
this.notifications =  remoteInterface<NotificationInterface>()
```
and then changing all the function usage from
```typescript
this.createNotif(...)
```
   to
```typescript
this.notifications.createNotification(...)
```

ShishKabab commented:
Wow, I didn't realize the scale of this. Because of this, I'd say that it's OK to set up the initial structure, and port the entire codebase generally. What were the challenges you faced implementing the initial proposal? That one didn't require the UI itself to be aware of the fact that it's a remote interface apart from a central setup once the refactor is complete (a next step would be to set these things up in top-level components, and as them down as needed), which is one goal I'm aiming for, because this separation may not exist once we re-use the code mobile for example.

As in the proposal, it's not necessary to organize these functions into classes yet. You could just do something like
```
setupRemoteFunctions({
  notifications: {
    createNotification: 'wherever it comes from'
  }
})
```

`setupRemoteFunctions` would just iterate through the object and register relevant functions (which can be accept the same args as remote functions with extra args, like `(info : {tabId}, params) => Promise<any>`) registered as `notifications:createNotification`.

But great you're already uncovering unclear/buggy parts of the codebase! The internal analytics are meant to be totally client-side for the client to act on, like showing useful notifications, but are not used yet as to my knowledge.

cdharris commented:
Okay thanks!

I think this looks a little closer now from the client side point of view of abstraction from the remote functions.

I've now changed a consuming client, i.e. `src/options/blacklist/actions.ts` from:
```typescript
import { NotificationInterface } from 'src/util/notification-types'
const createNotifRPC = remoteInterface<NotificationInterface>().createNotification
//...
createNotifRPC({ ... })
```
to
```typescript
import { remoteFunctions } from 'src/util/remote-functions'
const { notifications } = remoteFunctions
//...
notifications.createNotification({ ... })
```
So the client can now just use the function as is, it just needs to import it from the remote module, I assume that part is unavoidable, unless as you say, getting it from React context or something in the future.

I was previously a little confused about the registration of these interfaces, possibly over-complicating the dynamic type generics. For now I've now just typed the main remote registry (`RemoteFunctionsReg` in `src/util/remote-functions.ts:4`). But I would need to now also cover the typed remote function registration.

cdharris commented:
Some update on where I've got to with this.
- Updated documentation on `addBookmark` / `addPageBookmark` after feedback from Jon. There's some further refactoring to be done there outside of the scope of this task.

- Made `makeRemotelyCallable` optionally callable with a type (e.g.  `makeRemotelyCallable<BookmarkInterface>`) in order to check the functions provided implement that interface. So that the corresponding remote function calls `RunInBackground<BookmarkInterface>.func...` can be sure to match up correctly.

- Made a typesafe version of `remoteFunction(..)` namely `runInBackground<T>(..)`, so that we can iteratively refactor these remote functions (as there is quite a lot of them).

- Started refactoring the usage of remote functions:

The functions done are now behind in the interfaces `NotificationInterface` and `BookmarksInterface`, namely: `createNotification`, `addPageBookmark`, and `delPageBookmark`.

I've changed their client callsite usage from e.g:
```typescript
const createNotifRPC = remoteFunction('createNotification')
this.createNotif({ ... })
```
to
```typescript
import { notifications } from 'src/util/remote-functions-background'
notifications.createNotification({ ... })
```
(where the notifications import does:)
```typescript
export const notifications = runInBackground<NotificationInterface>()
```

- Removed ambiguity between the similar looking `remoteFunction` calls:

These either executed in the background or executed in a tab. I've hoped to make this clearer by naming them appropriately and forwarding them to the respective `remoteFunction` i.e.
from:
```typescript
remoteFunction('extractPageContent', { tabId })()
remoteFunction('extractPageContent', )({ tabId })
```
to:
```typescript
runInTab<PageAnalyzerInterface>(tabId).extractPageContent()
runInBackground<PageAnalyzerInterface>().extractPageContent({ tabId })
```

- Refactored all remote functions that run in tabs to this new usage. The nature of the higher order function means they are always called with a `tabId`, so the invocation looks a bit different than `runInBackground` and doesn't lend itself to being setup ahead of time. So I've kept the invocations similar, and just replaced the called instances to their less-ambiguous typesafe versions, i.e. from `remoteFunction('insertOrRemoveRibbon', { tabId: this.id })()
` to `runInTab<RibbonInteractionsInterface>( this.id ).insertOrRemoveRibbon()`


This all might not be perfect yet but hopefully it's getting better.

ShishKabab commented:
Looks very nice Chris! As a small thing to be fixed, I saw that `SearchBackground.remoteFunctions` doesn't have any type yet.

Also, it'd be great if you could do some re-organization. I'd move everything that has to do with tabs to something like `src/ipc` or `src/rpc`. Later on, I'd like to re-organize the entire code layout a bit to separate features from sub-systems or layers, since it's getting a bit too big and messy to handle. `src/util/notif*` would also be nice to have in another place, because it's not a util, but browser interface abstraction, which probably should live in `src/notifications/` somewhere.

Then, as a final step I'd make some over-arching types and/or functions containing the different remote interfaces. The idea is to in the background be sure that we have really registered all modules:
```
setupRemoteFunctionsImplementations({
    notifications: { createNotification }, // commenting out this line should give a compile-time error
    bookmarks: {
        addPageBookmark: search.remoteFunctions.addPageBookmark,
        delPageBookmark: search.remoteFunctions.delPageBookmark,
    },
} as RemoteFunctionImplementations)
```

From the consuming side, it'd be nice not to have to set up each module seperately, but instead have again a proxy which is bound to the interface:
```
const remoteFunctions : RemoteFunctionCalls = setupRemoteFunctionsCalls()
export default remoteFunctions
```

`RemoteFunctionImplementations` should be a type mappable from `RemoteFunctionCalls`.

In that way, we have everything type step to the last step  :)

In a later stage, we can also name `remoteFunctions` to `services`, which removes the whole assumption of that it's being ran somewhere else, instead talking about different components of a system.

cdharris commented:
Yep that all makes sense to me too, thanks for the feedback!

I was deferring that final glue interface `RemoteFunctionImplementations` but you're right to suggest including it. Thanks for the concrete example. 

I'll make those final revisions.

cdharris commented:
Couple of commits and thoughts following that feedback:

- Create that final over-arching interface as suggested to be sure we've registered all that's used.
- tested & manually tested - seems good!
- added notes to `refactoring.md`

I did start re-organising the notifications as suggested, but found some more inconsistencies so documented them in the refactoring doc and differed till then.

Likewise I wasn't completely sure which bit of the tabs functionality to re-organise, there's a lot of usage from within react components. Need some more clarification on the direction and scope here - have added it as an item to the refactoring doc for now.

RE the consuming interface, I've kept the individual exports for now as discussed so the client side doesn't need to de-reference them from an imported 'remoteFunctions' object and instead can use them more like the real things.

ShishKabab commented on src/search/bookmarks.ts:
```diff
@@ -14,6 +14,7 @@ export const addBookmark = (getDb: DBGet, tabManager: TabManager) => async ({
     timestamp?: number
     tabId?: number
 }) => {
+    // url = normalizeUrl(url);
```
Is this intentional?

cdharris replied:
Good catch. Left over from refactoring. Not needed. (this whole function is not currently used just serves as a reference of the new style, documented in refactoring.md)

ShishKabab commented on src/search/background/index.ts:
```diff
@@ -13,13 +13,15 @@ import {
 
 export default class SearchBackground {
     storage: SearchStorage
     private backend
```
Still needs type

cdharris replied:
Very true! Though this isn't something I've added. It stands as another set of remote function implementation refactoring needed. (detailed in refactoring.md). I've added a note specifically in the code to communicate that.

ShishKabab commented:
Very cool Chris, looks all good! Aside from those two little things to look at, I'd say this is work is finished. Congrats on this complex PR requiring quite some code diving!

cdharris commented:
Thanks Vincent! Hopefully this is done for now :)

ShishKabab commented:
Whenever you want @blackforestboi you can merge this together with a bunch of other features and test  :)

blackforestboi commented:
@cdharris What are the features affected by this? (so I can keep an eye open!)

cdharris commented:
@blackforestboi The features refactored in this PR include Bookmarks (adding and removing), Notifications, and things that RunInTab: Inserting/Removing the Ribbon, inserting/removing the tooltip, extracting page content. 

Hopefully it's all good :)

blackforestboi commented:
Its all working! :) Wonderful!
Will push a new update tomorrow!
